src/compiler/checker.ts

```diff
@@ -19789,6 +19789,36 @@ namespace ts {
             }
         }
 
+        function checkJSDocAugmentsTag(node: JSDocAugmentsTag): void {
+            const cls = getNodeCommentedByJSDocTag(node);
+            if (!isClassDeclaration(cls) && !isClassExpression(cls)) {
+                error(cls, Diagnostics.JSDoc_augments_tag_will_be_ignored_if_not_attached_to_a_class_declaration);
+                return;
+            }
+
+            const name = node.class.expression.kind === SyntaxKind.PropertyAccessExpression ? node.class.expression.name : node.class.expression;
+            const extend = getClassExtendsHeritageClauseElement(cls);
+            if (extend) {
+                const className = getClassName(extend.expression);
+                if (className && name.escapedText !== className.escapedText) {
+                    error(name, Diagnostics.JSDoc_augments_tag_declares_to_extend_0_but_actual_class_augments_1,
+                        unescapeLeadingUnderscores(name.escapedText),
+                        unescapeLeadingUnderscores(className.escapedText));
+                }
+            }
+        }
+
+        function getClassName(node: Expression): Identifier | undefined {

```

Comment:
This feels like it already exists. Either way, `getClassName` is too vague+specific a name for something that treats an Expression like a `QualifiedName` and then gets the identifier at the bottom. `getIdentifierFromExpressionAsQualifiedName`? That's horrible in a different way. I'm not sure what the best name is.